src/uu/tail/src/args.rs

```diff
@@ -358,8 +409,37 @@ fn parse_num(src: &str) -> Result<Signum, ParseSizeError> {
 }
 
 pub fn parse_args(args: impl uucore::Args) -> UResult<Settings> {
-    let matches = uu_app().try_get_matches_from(arg_iterate(args)?)?;
-    Settings::from(&matches)
+    let args_vec: Vec<OsString> = args.collect();
+    let clap_args = uu_app().try_get_matches_from(args_vec.clone());
+    let clap_result = match clap_args {
+        Ok(matches) => Ok(Settings::from(&matches)?),
+        Err(err) => Err(err.into()),
+    };
+
+    // clap isn't able to handle obsolete syntax.
+    // therefore, we want to check further for obsolete arguments.
+    // argv[0] is always present, argv[1] might be obsolete arguments
+    // argv[2] might contain an input file, argv[3] isn't allowed in obsolete mode
+    if args_vec.len() != 2 && args_vec.len() != 3 {
+        return clap_result;
+    }
+
+    // there are cases of ambiguity between clap's parsing and obsolete syntax parsing.
+    // e.g. "tail -c 5", which is a valid flag (handled by clap),
+    // but could also be interpreted as valid obsolete syntax (tail -c on file '5').
+    // in this case, the clap result should be returned immediately.
+    // clap fails for unknown flags, but valid obsolete arguments (e.g. tail -l; tail -10c).
+    // clap succeeds for obsolete arguments starting with '+', but misinterprets them as
+    // input files (e.g. tail +f).
+    // in these cases, the argv[1] should be checked further.
```

Comment:
Nice, I like that you explained more cases than I did, but let's bring back some of the formatting I had in my version and fix some of the punctuation. I tried to combine our approaches (again, feel free to improve further):

```rust
// At this point, there are a few possible cases:
//
//    1. clap has succeeded and the arguments would be invalid for the obsolete syntax.
//    2. The case of `tail -c 5` is ambiguous. clap parses this as `tail -c5`,
//       but it could also be interpreted as valid obsolete syntax (tail -c on file '5').
//       GNU chooses to interpret this as `tail -c5`, like clap.
//    3. `tail -f foo` is also ambiguous, but has the same effect in both cases. We can safely
//        use the clap result here.
//    4. clap succeeded for obsolete arguments starting with '+', but misinterprets them as
//       input files (e.g. tail +f).
//    5. clap failed because of unknown flags, but possibly valid obsolete arguments
//        (e.g. tail -l; tail -10c).
//
// In cases 4 & 5, we want to try parsing the obsolete arguments, which corresponds to
// checking whether clap succeeded or the first argument starts with '+'.
```

Reply:
Thanks for helping out with the format!